include/uk/arch/paging.h

```diff
@@ -0,0 +1,436 @@
+#include <uk/essentials.h>
+
+/* Cause a compiler error if not declared */
+typedef __pte_t __pte_t;	/* page table entry */
+
+/**
+ * __VADDR_INV / __PADDR_INV definitions
+ *
+ * A value that can be used to express an invalid virtual / physical address.
+ *
+ * Note: The value must be aligned to the largest supported page size!
+ */
+
+#ifndef __VADDR_INV
+#error Incomplete definition of virtual address type
+#endif
+
+#ifndef __PADDR_INV
+#error Incomplete definition of physical address type
+#endif
+
+/* Just an alias to express an arbitrary virtual / physical address */
+#define __VADDR_ANY			__VADDR_INV
+#define __PADDR_ANY			__PADDR_INV
+
+/**
+ * PT_Lx_IDX(vaddr, lvl)
+ *
+ * @param vaddr a virtual address
+ * @param lvl a page table level [0..PT_LEVELS - 1]
+ *
+ * @return the index of the page table entry (starting from 0) in the page
+ *    table at the given level that is used to translate the virtual address
+ */
+#ifndef PT_Lx_IDX
+unsigned int PT_Lx_IDX(__vaddr_t vaddr, unsigned int lvl);
+#endif
+
+/**
+ * PT_Lx_PTES(lvl) macro
+ *
+ * NOTE: Must be compile-time resolvable
+ *
+ * @param lvl a page table level [0..PT_LEVELS - 1]
+ *
+ * @return the number of page table entries per page table at the given level
+ */
+#ifndef PT_Lx_PTES
+#error PT_Lx_PTES not defined
+#endif
+
+/**
+ * PT_Lx_PTE_PRESENT(pte, lvl)
+ *
+ * @param pte a page table entry from a page table at the given level
+ * @param lvl a page table level [0..PT_LEVELS - 1]
+ *
+ * @return a non-zero value if the page table entry is present, that is the HW
+ *    would use it for address translation. The function must return 0 for
+ *    PT_Lx_PTE_INVALID(lvl).
+ */
+#ifndef PT_Lx_PTE_PRESENT
+int PT_Lx_PTE_PRESENT(__pte_t pte, unsigned int lvl);
+#endif
+
+/**
+ * PT_Lx_PTE_CLEAR_PRESENT(pte, lvl)
+ *
+ * @param pte a page table entry from a page table at the given level
+ * @param lvl a page table level [0..PT_LEVELS - 1]
+ *
+ * @return a modified version of the input PTE so that the HW will not use the
+ *    PTE for address translation. Any bits ignored by the HW in this invalid
+ *    state should remain untouched.
+ */
+#ifndef PT_Lx_PTE_CLEAR_PRESENT
+__pte_t PT_Lx_PTE_CLEAR_PRESENT(__pte_t pte, unsigned int lvl);
+#endif
+
+/**
+ * PT_Lx_PTE_INVALID(lvl)
+ *
+ * @param lvl a page table level [0..PT_LEVELS - 1]
+ *
+ * @return a PTE value that expresses an invalid state in a page table at the
+ *    given level where the HW will not use the page table entry to translate
+ *    a virtual address.
+ */
+#ifndef PT_Lx_PTE_INVALID
+__pte_t PT_Lx_PTE_INVALID(unsigned int level);
+#endif
+
+/**
+ * PT_Lx_PTE_PADDR(pte, lvl)
+ *
+ * @param pte a page table entry from a page table at the given level
+ * @param lvl a page table level [0..PT_LEVELS - 1]
+ *
+ * @return the physical address to which the given page table entry points
+ */
+#ifndef PT_Lx_PTE_PADDR
+__paddr_t PT_Lx_PTE_PADDR(__pte_t pte, unsigned int lvl);
+#endif
+
+/**
+ * PAGE_Lx_SHIFT(lvl)
```

Comment:
I found the name of this macro confusing at first, as the values returned are also the number of bits to shift obtain the level+1 bits in an address. Eg the order of L0 is 12, and the shift for L1 bits is 12 too.

Reply:
Don't quite get it 🙈 On x86 you have:
```C
PAGE_Lx_SHIFT(0) = 12;  /* 0 = PAGE_LEVEL */
PAGE_Lx_SHIFT(1) = 21;  /* 1 = PAGE_LARGE_LEVEL */
PAGE_Lx_SIZE(0) = 4 KiB;
PAGE_Lx_SIZE(1) = 2 MiB;
```